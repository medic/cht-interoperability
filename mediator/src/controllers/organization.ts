import axios from "axios";
import { FHIR } from "../../config";
import { logger } from "../../logger";

const { url: fhirUrl, username, password } = FHIR;

export async function createOrganization(organization: fhir5.Organization) {
  try {
    return await axios.post(`${fhirUrl}/Organization`, organization, {
      auth: { username, password },
    });
  } catch (error: any) {
    logger.error(error);
    return { status: error.status, data: { message: error.data } };
  }
}
