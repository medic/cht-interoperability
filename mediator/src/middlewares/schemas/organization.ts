import joi from 'joi';

const EndpointReferenceSchema = joi.object({
  reference: joi.string().regex(/Endpoint\/\S+/).required(),
});

// todo => add the requirement for an identifier
export const OrganizationSchema = joi.object({
  name: joi.array().length(1).required(),
  endpoint: joi.array().items(EndpointReferenceSchema).length(1).required(),
});
