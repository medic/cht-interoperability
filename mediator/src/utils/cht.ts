import axios from 'axios';
import { CHT } from '../../config';
import { generateBasicAuthUrl } from './url';
import https from 'https';
import path from 'path';

export async function createChtRecord(patientId: string) {
  const record = {
    _meta: {
      form: 'interop_follow_up',
    },
    patient_uuid: patientId,
  };
  const options = {
    httpsAgent: new https.Agent({
      rejectUnauthorized: false,
    }),
  };

  const chtApiUrl = generateChtRecordsApiUrl(CHT.url, CHT.username, CHT.password);

  return await axios.post(chtApiUrl, record, options);
}

export const generateChtRecordsApiUrl = (chtUrl: string, username: string, password: string) => {
  const endpoint = generateBasicAuthUrl(chtUrl, username, password);
  return path.join(endpoint, '/api/v2/records');
};
