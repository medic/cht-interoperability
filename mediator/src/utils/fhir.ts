export const VALID_GENDERS = ['male', 'female', 'other', 'unknown'] as const;

export interface IPatient {
  name: string;
  date_of_birth: string;
  _id: string;
  sex: typeof VALID_GENDERS[number];
}

export function generateFHIRPatientResource(patient: IPatient) {
  if (!patient.name) {
    throw new Error(`Invalid 'name' expected type of 'string' but recieved '${patient.name}'`);
  }
  
  const patientLastName = patient.name.split(' ').slice(-1);
  const birthDate = new Date(patient.date_of_birth);

  if (!isValidDate(birthDate)) {
    throw new RangeError("Invalid 'date_of_birth' range: received " + patient.date_of_birth);
  } else if (!patient._id) {
    throw new Error(`Invalid '_id' epxted type of 'string' or 'number' but recieved '${typeof patient._id}' with value '${patient._id}'`);
  }  else if (!patient.sex || !VALID_GENDERS.includes(patient.sex)) {
    throw new Error(`Invalid 'sex' expected 'male', 'female', 'other', 'unknown' but recieved '${patient.sex}'`);
  }

  const FHITPatientResource = {
    resourceType: 'Patient',
    id: patient._id,
    identifier: [
      {
        system: 'cht',
        value: patient._id
      }
    ],
    name: [
      {
        use: 'official',
        family: patientLastName,
        given: [patient.name]
      }
    ],
    gender: patient.sex,
    birthDate: birthDate.toISOString()
  };

  return FHITPatientResource;
}

export function isValidDate(d: Date) {
  return d instanceof Date && !isNaN(d as any);
}

export function generateFHIRSubscriptionResource(patientId: string, callbackUrl: string) {
  if (!patientId) {
    throw new Error(`Invalid patient id was expecting type of 'string' or 'number' but received '${typeof patientId}'`)
  } else if (!callbackUrl) {
    throw new Error(`Invalid 'callbackUrl' was expecting type of 'string' but recieved '${typeof callbackUrl}'`)
  }
  
  const FHIRSubscriptionResource = {
    resourceType: 'Subscription',
    id: patientId,
    status: 'requested',
    reason: 'Follow up request for patient',
    criteria: `Encounter?identifier=${patientId}`,
    channel: {
      type: 'rest-hook',
      endpoint: callbackUrl,
      payload: 'application/fhir+json',
      header: ['Content-Type: application/fhir+json']
    }
  };

  return FHIRSubscriptionResource;
}
