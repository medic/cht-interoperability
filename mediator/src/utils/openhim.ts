import { logger } from "../../logger";

export const registerMediatorCallback = (err?: string): void => {
  if (err) {
    throw new Error(`Mediator Registration Failed: Reason ${err}`);
  }

  logger.info('Successfully registered mediator.');
};
