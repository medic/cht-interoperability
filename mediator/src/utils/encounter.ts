import { CHTEncounterDoc } from '../controllers/encounter';

export function generateFHIREncounterResource(encounter: CHTEncounterDoc) {
  const FHIREncounterResource = {
    resourceType: 'Encounter',
    identifier: [
      {
        use: 'official',
        value: encounter.patient_id
      }
    ],
    status: 'finished',
    class: 'outpatient',
    type: [
      { text: 'Community health worker visit' }
    ],
    subject: encounter.patient_id,
    participant: [
      {
        type: [
          { text: 'Community health worker' }
        ]
      }
    ]
  };

  return FHIREncounterResource;
}
