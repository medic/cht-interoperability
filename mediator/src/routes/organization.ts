import { Router } from 'express';
import { validateBodyAgainst } from '../middlewares';
import { createFhirResource, validateFhirResource } from '../utils/fhir';
import { OrganizationSchema } from '../middlewares/schemas/organization';
import { requestHandler } from '../utils/request';

const router = Router();

const resourceType = 'Organization';

router.post(
  '/',
  validateBodyAgainst(validateFhirResource(resourceType), OrganizationSchema),
  requestHandler((req) => createFhirResource({ ...req.body, resourceType }))
);

export default router;
