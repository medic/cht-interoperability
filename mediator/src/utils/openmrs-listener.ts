import { chtEventEmitter, CHT_EVENTS } from './cht';
import { sendPatientToOpenMRS, sendEncounterToOpenMRS } from './openmrs_sync';
import { logger } from '../../logger';
import { createSubscription, fhirEventEmitter } from './fhir';
import { OPENMRS } from '../../config';

// Store active listeners to allow for deregistration
type PatientListener = (patient: fhir4.Patient) => Promise<void>;
type EncounterListener = (data: { encounter: fhir4.Encounter, references: fhir4.Resource[] }) => Promise<void>;
const activeListeners: { [key: string]: PatientListener | EncounterListener } = {};

function registerPatientListener() {
  if (activeListeners[CHT_EVENTS.PATIENT_CREATED]) {
    return; // Already registered
  }
  const listener = async (patient: fhir4.Patient) => {
    try {
      await sendPatientToOpenMRS(patient);
    } catch (error) {
      logger.error(`Error sending patient to OpenMRS: ${error}`);
    }
  };
  chtEventEmitter.on(CHT_EVENTS.PATIENT_CREATED, listener);
  activeListeners[CHT_EVENTS.PATIENT_CREATED] = listener;
  logger.info('Patient listener registered');
}

function registerEncounterListener() {
  if (activeListeners[CHT_EVENTS.ENCOUNTER_CREATED]) {
    return; // Already registered
  }
  const listener = async (data: { 
    encounter: fhir4.Encounter, 
    references: fhir4.Resource[] 
  }) => {
    try {
      await sendEncounterToOpenMRS(data.encounter, data.references);
    } catch (error) {
      logger.error(`Error sending encounter to OpenMRS: ${error}`);
    }
  };
  chtEventEmitter.on(CHT_EVENTS.ENCOUNTER_CREATED, listener);
  activeListeners[CHT_EVENTS.ENCOUNTER_CREATED] = listener;
  logger.info('Encounter listener registered');
}

function deregisterListener(eventName: string) {
  if (activeListeners[eventName]) {
    chtEventEmitter.off(eventName, activeListeners[eventName]);
    delete activeListeners[eventName];
    logger.info(`Deregistered listener for ${eventName}`);
  }
}

function handleActiveSubscription(subscription: fhir4.Subscription) {
  if (subscription.criteria.startsWith('Patient')) {
    registerPatientListener();
  } else if (subscription.criteria.startsWith('Encounter')) {
    registerEncounterListener();
  }
}

// Handle subscription status changes
fhirEventEmitter.on('resourceCreated', (resource: fhir4.Resource) => {
  if (resource.resourceType === 'Subscription' && (resource as fhir4.Subscription).status === 'active') {
    handleActiveSubscription(resource as fhir4.Subscription);
  }
});

fhirEventEmitter.on('resourceUpdated', (resource: fhir4.Resource) => {
  if (resource.resourceType === 'Subscription') {
    const subscription = resource as fhir4.Subscription;
    if (subscription.status === 'active') {
      handleActiveSubscription(subscription);
    } else {
      if (subscription.criteria.startsWith('Patient')) {
        deregisterListener(CHT_EVENTS.PATIENT_CREATED);
      } else if (subscription.criteria.startsWith('Encounter')) {
        deregisterListener(CHT_EVENTS.ENCOUNTER_CREATED);
      }
    }
  }
});

export async function registerOpenMRSListeners() {
  try {
    // Create subscriptions for patient and encounter events
    await createSubscription(
      'Patient?',
      `${OPENMRS.url}/Patient/`,
      ['Content-Type: application/fhir+json']
    );
    
    await createSubscription(
      'Encounter',
      `${OPENMRS.url}/Encounter/`,
      ['Content-Type: application/fhir+json']
    );

    logger.info('OpenMRS FHIR subscriptions created successfully');
  } catch (error) {
    logger.error('Error creating OpenMRS FHIR subscriptions:', error);
  }
}
