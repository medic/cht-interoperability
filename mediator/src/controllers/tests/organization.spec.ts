import axios from "axios";
import { logger } from "../../../logger";
import { createOrganization } from "../organization";
import { OrganizationFactory } from "../../middlewares/schemas/tests/utils";

jest.mock("../../../logger");
jest.mock("axios");
const mockAxios = axios as jest.Mocked<typeof axios>;

describe("Organization controllers", () => {
  const organization: fhir4.Organization = OrganizationFactory.build();

  describe("createOrganization", () => {
    it("creates a organization when given a valid organization resource", async () => {
      const data = { status: 200, data: {} };

      mockAxios.post.mockResolvedValueOnce(data);

      const res = await createOrganization(organization);

      expect(res.data).toBe(data.data);
      expect(res.status).toEqual(data.status);
      expect(mockAxios.post).toHaveBeenCalled();
      expect(mockAxios.post.mock.calls[0][0]).toContain("/fhir/Organization");
      expect(mockAxios.post.mock.calls[0][1]).toStrictEqual(organization);
      expect(logger.error).not.toHaveBeenCalled();
    });

    it("resolves error when the organization creation fails", async () => {
      const data = { status: 400, data: {} };

      mockAxios.post.mockRejectedValueOnce(data);

      const res = await createOrganization(organization);

      expect(res.status).toBe(data.status);
      expect(logger.error).toHaveBeenCalled();
      expect(mockAxios.post).toHaveBeenCalled();
      expect(res.data).toBe(res.data);
    });
  });
});
