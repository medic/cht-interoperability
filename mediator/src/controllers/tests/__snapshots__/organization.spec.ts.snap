// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Organization controllers createOrganization creates a organization when given a valid organization resource 1`] = `
[MockFunction] {
  "calls": [
    [
      "http://openhim-core:5001/fhir/Organization",
      {
        "resourceType": "Organization",
      },
      {
        "auth": {
          "password": "interop-password",
          "username": "interop-client",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
